### PR TITLE
New invitation endpoint

### PR DESCRIPTION
**Pull request by joeatwork**
An alternative to https://github.com/coreos/dex/pull/157 - allow newly invited users to simultaneously verify their email addresses and reset their passwords.

### CODE REVIEW

**Comment by joeatwork** [comment]
Discussed a simpler approach with @bobbyrullo 

- Invitation endpoint verifies email address, produces a second, pure PasswordChange token, redirects the user.

- Consider adding a method to the InvitationToken that can produce a PasswordChange token, preserving everything it needs

That means you can get rid of handlePOST and a bunch of the new test cases

File: server/password_test.go [hunk 1]
```diff
@@ -355,9 +361,35 @@ func TestSendResetPasswordEmailHandler(t *testing.T) {
 	}
 }
 
+type resetPasswordTestCase struct {
```
Comment (bobbyrullo) [inline server/password_test.go]: Why define this outside of testResetPasswordHandler?

File: server/password_test.go [hunk 2]
```diff
@@ -355,9 +361,35 @@ func TestSendResetPasswordEmailHandler(t *testing.T) {
...
+
+	wantFormValues *url.Values
+	wantCode       int
+	wantPassword   string
+}
+
 func TestResetPasswordHandler(t *testing.T) {
-	makeToken := func(userID, password, clientID string, callback url.URL, expires time.Duration, signer jose.Signer) string {
-		pr := user.NewPasswordReset(user.User{ID: "ID-1"},
+	rpwUserID := "ID-1"
```
Comment (bobbyrullo) [inline server/password_test.go]: why all these changes? I don't think it's necessary to prefix all this stuff with "rpw" and "pw" - these variables are local to this file anyhow. In addition, it makes this particular commit larger and harder to read - it's difficult to see where the real changes are.

File: user/email/email.go [hunk 3]
```diff
@@ -46,76 +48,103 @@ func NewUserEmailer(ur user.UserRepo,
...
+	}
+
+	resetURL := u.invitationURL
+	q := resetURL.Query()
+	q.Set("token", token)
+	resetURL.RawQuery = q.Encode()
 
 	if u.emailer != nil {
-		err = u.emailer.SendMail(u.fromAddress, subj, tmplName,
+		err = u.emailer.SendMail(u.fromAddress, "Reset Your Password", "password-reset",
```
Comment (bobbyrullo) [inline user/email/email.go]: Shouldn't this be "Activate your account" and use the invitation template?

File: user/api/api.go [hunk 4]
```diff
@@ -88,7 +88,6 @@ type UsersAPI struct {
 }
 
 type Emailer interface {
-	SendResetPasswordEmail(string, url.URL, string) (*url.URL, error)
```
Comment (bobbyrullo) [inline user/api/api.go]: !!!! why?

File: server/password_test.go (hunk 1 above)
Comment (joeatwork) [inline server/password_test.go]: Made obsolete in a162744

File: server/password_test.go (hunk 2 above)
Comment (joeatwork) [inline server/password_test.go]: Made obsolete in a162744

File: user/email/email.go (hunk 3 above)
Comment (joeatwork) [inline user/email/email.go]: Fixed in 93df33f

**Pull request merged**